rust/lance-io/src/object_writer.rs

```diff
@@ -58,44 +274,185 @@ impl ObjectWriter {
     }
 }
 
-#[async_trait]
-impl Writer for ObjectWriter {
-    async fn tell(&mut self) -> Result<usize> {
-        Ok(self.cursor)
+impl Drop for ObjectWriter {
+    fn drop(&mut self) {
+        // If there is a multipart upload started but not finished, we should abort it.
+        if matches!(self.state, UploadState::InProgress { .. }) {
+            // Take ownership of the state.
+            let state = std::mem::replace(&mut self.state, UploadState::Done);
+            if let UploadState::InProgress { mut upload, .. } = state {
+                tokio::task::spawn(async move {
+                    let _ = upload.abort().await;
+                });
+            }
+        }
     }
 }
```

Comment:
This is what ensures partial uploads are cleaned up now. `MultipartId` is no longer exposed in the `object-store` API.